p4-16/psa/PSA.mdk

```diff
@@ -573,15 +580,15 @@ A similar FIFO property is expected to hold for multicast packets,
 i.e. those that follow the "Make 0 or more copies" path in the
 pseudocode above.  By a short period of time after the control plane
 has stopped modifying the set of members of multicast groups, they
-should be 'stable', and all packets with the same `(ingress_port,
-ostd.multicast_group, ostd.class_of_service, egress_port)` should pass
-through the system in FIFO order, and be processed in the `Egress`
-control block in the same relative order that they were processed in
-the `Ingress` control block.
+should be 'stable', and all packets with the same `(ingress_port,`
+`ostd.multicast_group,` `ostd.class_of_service,` `egress_port)` should
+pass through the system in FIFO order, and be processed in the
+`Egress` control block in the same relative order that they were
+processed in the `Ingress` control block.
```

Comment:
Do we really want to imply such ordering guarantees? Is it important for portability to rely on a FIFO impl?

Reply:
As currently written, this is not a requirement.  We can leave it out if it is too restrictive on implementations, but note that there are likely to be some P4 programs that do updates on registers in ingress and egress that might implicitly _assume_ such ordering properties, and fail on devices that don't.

Reply:
Commercial switches typically implement these ordering restrictions as packets are transmitted on the output ports, at least.  Whether or not they implement that by making the same restriction on the relative order that packets pass through their internal processing pipelines, or by allowing out of order processing, but then reorder before transmitting the packets on the wire, is an implementation detail.

If a commercial top-of-rack switch did _not_ implement this port-in-to-port-out ordering constraint, several UDP-based applications would break, and several (but not all) TCP implementations in operating systems would suffer factors of 10 or larger drops in application throughput, because they misinterpret too much packet reordering in the network as packet drops, thus a sign of network congestion, and reduce their TCP transmission window.  That is why LAGs and ECMP are typically selected among using a hash of packet header fields, not packet-by-packet round robin.